packages/components/src/utils/react.ts

```diff
@@ -0,0 +1,244 @@
+/**
+ * WordPress dependencies
+ */
+import {
+	useRef,
+	useInsertionEffect,
+	useCallback,
+	useEffect,
+	useState,
+} from '@wordpress/element';
+
+/**
+ * Any function.
+ */
+export type AnyFunction = ( ...args: any ) => any;
+
+/**
+ * Creates a stable callback function that has access to the latest state and
+ * can be used within event handlers and effect callbacks. Throws when used in
+ * the render phase.
+ *
+ * @example
+ *
+ * ```tsx
+ * function Component(props) {
+ *   const onClick = useEvent(props.onClick);
+ *   React.useEffect(() => {}, [onClick]);
+ * }
+ * ```
+ */
+export function useEvent< T extends AnyFunction >( callback?: T ) {
```

Comment:
cc @youknowriad for now we're keeping this hook internal to components package. But if you think it would be useful, we could also move it to `@wordpress/compose`

Reply:
Definitely something that looks useful. I thought at some point that React was working on providing it no?

Reply:
They were, but from what I understand the release of the hook has been put on hold. We could add it anyway, and if it ever gets officially released in react, we can move to the official one

Reply:
Took note of this! Resolving.

Reply:
Update: https://github.com/WordPress/gutenberg/pull/64943